langchain/retrievers/perfect_retriever.py

```diff
@@ -0,0 +1,220 @@
+
+pinecone_example = """EXAMPLE
+----------------------------
+
+Here is an example:
+
+DOCUMENT STORE DESCRIPTION: News headlines from around the world
+METADATA FIELDS: {{{{
+    "year": "The year the headline was published", 
+    "country": The country of origin of the news media outlet, 
+    "source": "The name of the news media outlet"
+}}}}
+QUESTION: What was the sentiment of Chinese and Taiwanese media outlets regarding the 2024 trade deal between America and India?
+DOCUMENT STORE QUERY:
+```json
+{{{{
+    "search_string": "Trade deal between America and India",
+    "metadata_filter": {{{{
+        "country": {{{{
+            "$in": ["China", "Taiwan"]
+        }}}},
+        "year": {{{{
+            "$eq": 2024
+        }}}}
+    }}}}
+}}}}
+```"""
+
+self_query_prompt = """INSTRUCTIONS
+----------------------------
+
+You have access to a store of documents. Each document contains text and a key-value store of associated metadata. Given a user question, your job is to come up with a fully formed query to the document store that will return the most relevant documents.
+
+A document store query consists of two components: a search string and a metadata filter. The search string is compared to the text contents of the stored documents. The metadata filter is used to filter out documents whose metadata does not match the given criteria.
+
+
+{format_instructions}
+
+
+{example}
+
+Begin!
+
+DOCUMENT STORE DESCRIPTION: {{docstore_description}}
+METADATA FIELDS: {{metadata_fields}}
+QUESTION: {{{{question}}}}
+DOCUMENT STORE QUERY:""".format(
+    format_instructions=pinecone_format_instructions, example=pinecone_example
+)
+
+
+class VectorStoreQueryOutputParser(BaseOutputParser):
+    def get_format_instructions(self) -> str:
+        return pinecone_format_instructions
+
+    def parse(self, text: str) -> Any:
+        cleaned_output = text.strip()
+        if "```json" in cleaned_output:
+            _, cleaned_output = cleaned_output.split("```json")
+        if "```" in cleaned_output:
+            cleaned_output, _ = cleaned_output.split("```")
+        if cleaned_output.startswith("```json"):
+            cleaned_output = cleaned_output[len("```json") :]
+        if cleaned_output.startswith("```"):
+            cleaned_output = cleaned_output[len("```") :]
+        if cleaned_output.endswith("```"):
+            cleaned_output = cleaned_output[: -len("```")]
+        cleaned_output = cleaned_output.replace("`", "")
+        cleaned_output = cleaned_output.strip()
+        response = json.loads(cleaned_output)
+        return {
+            "query": response["search_string"],
+            "filter": response["metadata_filter"],
+        }
+
+
+class VectorStoreExtendedInfo(VectorStoreInfo):
+    metadata_field_descriptions: Dict[str, str]
+
+
+class VectorStoreSelfQueryRetriever(VectorStoreRetriever):
+    """"""
+
+    llm_chain: LLMChain
+    """"""
+
+    def get_relevant_documents(self, query: str) -> List[Document]:
```

Comment:
This method is completely overloaded without invocation of super (usage of inherited implementation). 

Does it make sense to change the parent from which we inherit to be a more basic interface or else rely more on a compositional pattern?

For testing / debugging purposes, is there a need to allow a user to specify query/filters and escape re-interpretation by the LLM?

Reply:
yea could maybe factor out relevant parts of VectorStoreRetriever into abstract BaseVectorStoreRetriever and inherit from that

my sense is the main functionality to test/debug here *is* the query/filter generation. if you just wanted to test what's retrieved for a given query/filter you could test the vectorstore directly

Reply:
and if you wanted to test/debug query/filter generation you could test the LLMChain directly